src/order/symm_diff.lean

```diff
@@ -0,0 +1,206 @@
+/-
+Copyright (c) 2021 Bryan Gin-ge Chen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz, Bryan Gin-ge Chen
+-/
+
+import order.boolean_algebra
+
+/-!
+# Symmetric difference
+
+The symmetric difference or disjunctive union of sets `A` and `B` is the set of elements that are
+in either `A` or `B` but not both. Translated into propositions, the symmetric difference is `xor`.
+
+The symmetric difference operator (`symm_diff`) is defined in this file for any type with `⊔` and
+`\` via the formula `(A \ B) ⊔ (B \ A)`, however the theorems proved about it only hold for
+`generalized_boolean_algebra`s and `boolean_algebra`s.
+
+The symmetric difference is the addition operator in the Boolean ring structure on Boolean algebras.
+
+## Main declarations
+
+* `symm_diff`: the symmetric difference operator, defined as `(A \ B) ⊔ (B \ A)`
+
+In generalized Boolean algebras, the symmetric difference operator is:
+
+* `symm_diff_comm`: commutative, and
+* `symm_diff_assoc`: associative.
+
+## Notations
+
+* `a Δ b`: `symm_diff a b`
+
+## References
+
+The proof of associativity follows the note "Associativity of the Symmetric Difference of Sets: A
+Proof from the Book" by John McCuan:
+
+* <https://people.math.gatech.edu/~mccuan/courses/4317/symmetricdifference.pdf>
+
+## Tags
+boolean ring, generalized boolean algebra, boolean algebra, symmetric differences
+-/
+
+/-- The symmetric difference operator on a type with `⊔` and `\` is `(A \ B) ⊔ (B \ A)`. -/
+def symm_diff {α : Type*} [has_sup α] [has_sdiff α] (A B : α) : α := (A \ B) ⊔ (B \ A)
+
+infix ` Δ `:100 := symm_diff
+
+lemma symm_diff_def {α : Type*} [has_sup α] [has_sdiff α] (A B : α) :
+  A Δ B = (A \ B) ⊔ (B \ A) :=
+rfl
+
+lemma symm_diff_eq_xor (p q : Prop) : p Δ q = xor p q := rfl
+
+section generalized_boolean_algebra
+variables {α : Type*} [generalized_boolean_algebra α] (a b c : α)
+
+lemma symm_diff_comm : a Δ b = b Δ a := by simp only [(Δ), sup_comm]
+
+instance symm_diff_is_comm : is_commutative α (Δ) := ⟨symm_diff_comm⟩
+
+@[simp] lemma symm_diff_self : a Δ a = ⊥ := by rw [(Δ), sup_idem, sdiff_self]
+@[simp] lemma symm_diff_bot : a Δ ⊥ = a := by rw [(Δ), sdiff_bot, bot_sdiff, sup_bot_eq]
+@[simp] lemma bot_symm_diff : ⊥ Δ a = a := by rw [symm_diff_comm, symm_diff_bot]
+
+lemma symm_diff_eq_sup_sdiff_inf : a Δ b = (a ⊔ b) \ (a ⊓ b) :=
+by simp [sup_sdiff, sdiff_inf, sup_comm, (Δ)]
+
+lemma disjoint_symm_diff_inf : disjoint (a Δ b) (a ⊓ b) :=
+begin
+  rw [symm_diff_eq_sup_sdiff_inf],
+  exact disjoint_sdiff.symm,
+end
+
+lemma symm_diff_le_sup : a Δ b ≤ a ⊔ b := by { rw symm_diff_eq_sup_sdiff_inf, exact sdiff_le }
+
+lemma sdiff_symm_diff : c \ (a Δ b) = (c ⊓ a ⊓ b) ⊔ ((c \ a) ⊓ (c \ b)) :=
+by simp only [(Δ), sdiff_sdiff_sup_sdiff']
+
+lemma sdiff_symm_diff' : c \ (a Δ b) = (c \ (a ⊔ b)) ⊔ (c ⊓ a ⊓ b) :=
```

Comment:
Rewriting by `sup_comm` seems like a distraction here:
```suggestion
lemma sdiff_symm_diff' : c \ (a Δ b) = (c ⊓ a ⊓ b) ⊔ (c \ (a ⊔ b)) :=
```
This makes it easier to spot the difference with the lemma above too.